flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java

```diff
@@ -405,54 +404,22 @@ public void addMasterState(MasterState state) {
 	// ------------------------------------------------------------------------
 
 	/**
-	 * Aborts a checkpoint because it expired (took too long).
+	 * Aborts a checkpoint with abort reason and cause.
 	 */
-	public void abortExpired() {
+	public void abort(CheckpointAbortReason reason, Throwable cause) {
 		try {
-			Exception cause = new Exception("Checkpoint expired before completing");
-			onCompletionPromise.completeExceptionally(cause);
-			reportFailedCheckpoint(cause);
+			onCompletionPromise.complete(new CheckpointExecuteResult(reason, cause));
```

Comment:
I found this a bit incosistent: why having a class that represents both, successful result and failure, when we could just `completeExceptionally` with a new type of maybe `CheckpointExecutionException` that carries the `reason` field. It seems that the client to `onCompletionPromise` has to handle exceptions anyways, see line 291 `onCompletionPromise.completeExceptionally(t);`

Reply:
As far as I can see `CheckpointExecutionException` is always used in the context a `CompletableFuture` that already is able to represent success and exceptional case.

Reply:
@StefanRRichter The `CheckpointExecuteResult` used to represent the execution result of Checkpoint just like `CheckpointTriggerResult` which means trigger result of Checkpoint. They would be both used in `CheckpointFailure` to counter the failure number(further work, step two, see : https://docs.google.com/document/d/1ce7RtecuTxcVUJlnU44hzcO2Dwq9g4Oyd8_biy94hJc/edit?usp=sharing).